service_container/service_locators.rst

```diff
@@ -0,0 +1,223 @@
+handlers could be to inject the whole dependency injection container::
+
+        use Symfony\Component\DependencyInjection\ContainerInterface;
+
+        class CommandBus
+        {
+            private $container;
+
+            public function __construct(ContainerInterface $container)
+            {
+                $this->container = $container;
+            }
+
+            public function handle(Command $command)
+            {
+                $commandClass = get_class($command)
+
+                if ($this->container->has($commandClass)) {
+                    $handler = $this->container->get($commandClass);
+
+                    return $handler->handle($command);
+                }
+            }
+        }
+
+However, injecting the entire container is discouraged because it gives too
+broad access to existing services and it hides the actual dependencies of the
+services.
+
+**Service Locators** are intended to solve this problem by giving access to a
+set of predefined services while instantiating them only when actually needed.
+
+Defining a Service Locator
+--------------------------
+
+First, define a new service for the service locator. Use its ``arguments``
+option to include as many services as needed to it and add the
+``container.service_locator`` tag to turn it into a service locator:
+
+.. configuration-block::
+
+    .. code-block:: yaml
+
+        services:
+
+            app.command_handler_locator:
+                class: Symfony\Component\DependencyInjection\ServiceLocator
+                arguments:
+                    AppBundle\FooCommand: '@app.command_handler.foo'
+                    AppBundle\BarCommand: '@app.command_handler.bar'
+                tags: ['container.service_locator']
+
+    .. code-block:: xml
+
+        <?xml version="1.0" encoding="UTF-8" ?>
+        <container xmlns="http://symfony.com/schema/dic/services"
+            xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+            xsi:schemaLocation="http://symfony.com/schema/dic/services http://symfony.com/schema/dic/services/services-1.0.xsd">
+
+            <services>
+
+                <service id="app.command_handler_locator" class="Symfony\Component\DependencyInjection\ServiceLocator">
+                    <argument key="AppBundle\FooCommand" type="service" id="app.command_handler.foo" />
+                    <argument key="AppBundle\BarCommand" type="service" id="app.command_handler.bar" />
+                    <tag name="container.service_locator" />
+                </service>
+
+            </services>
+        </container>
+
+    .. code-block:: php
+
+        use Symfony\Component\DependencyInjection\ServiceLocator;
+        use Symfony\Component\DependencyInjection\Reference;
+
+        //...
+
+        $container
+            ->register('app.command_handler_locator', ServiceLocator::class)
+            ->addTag('container.service_locator')
+            ->setArguments(array(
```

Comment:
I recently implemented a service locator [for API Platform](https://github.com/meyerbaptiste/core/commit/f2bc99f96d6d9834e1e7697f5e85a4f8c5c118a0#diff-e05d209a4d4088891bbb8428a57226aaR48) and the syntax described here is currently not the correct one.

According to [src/Symfony/Component/DependencyInjection/Compiler/ServiceLocatorTagPass.php#L40-L42](https://github.com/symfony/symfony/blob/cc5e582dcf86819ab034dcdfb5ce73f299a8aa08/src/Symfony/Component/DependencyInjection/Compiler/ServiceLocatorTagPass.php#L40-L42), this should be:
```php
->setArguments([
    [
        'AppBundle\FooCommand' => new Reference('app.command_handler.foo'),
        'AppBundle\BarCommand' => new Reference('app.command_handler.bar'),
    ],
])
```
Same for the YAML/XML config.